server/src/main/java/io/druid/segment/realtime/firehose/SqlFirehoseFactory.java

```diff
@@ -0,0 +1,164 @@
+      @JsonProperty("fetchTimeout") Long fetchTimeout,
+      @JsonProperty("foldCase") Boolean foldCase,
+      @JacksonInject SQLMetadataConnector sqlMetadataConnector,
+      @JacksonInject @Smile ObjectMapper objectMapper
+  )
+  {
+    super(
+        maxCacheCapacityBytes,
+        maxFetchCapacityBytes,
+        prefetchTriggerBytes,
+        fetchTimeout,
+        objectMapper
+    );
+
+    Preconditions.checkArgument(sqls.size() > 0, "No SQL queries provided");
+
+    this.sqls = sqls;
+    this.objectMapper = objectMapper;
+    this.sqlMetadataConnector = sqlMetadataConnector;
+    this.foldCase = (foldCase == null ? false : true);
+  }
+
+  @Override
+  protected InputStream openObjectStream(String object, File fileName) throws IOException
+  {
+    Preconditions.checkNotNull(sqlMetadataConnector, "SQL Metadata Connector not configured!");
+    try (FileOutputStream fos = new FileOutputStream(fileName)) {
+      sqlMetadataConnector.retryWithHandle(
+          (handle) -> {
+            ResultIterator<Map<String, Object>> resultIterator = handle.createQuery(
+                object
+            ).map(
+                (index, r, ctx) -> {
+                  Map<String, Object> resultRow = foldCase ? new CaseFoldedMap() : new HashMap<>();
+                  ResultSetMetaData resultMetadata;
+                  try {
+                    resultMetadata = r.getMetaData();
+                  }
+                  catch (SQLException e) {
+                    throw new ResultSetException("Unable to obtain metadata from result set", e, ctx);
+                  }
+                  try {
+                    for (int i = 1; i <= resultMetadata.getColumnCount(); i++) {
+                      String key = resultMetadata.getColumnName(i);
+                      String alias = resultMetadata.getColumnLabel(i);
+                      Object value = r.getObject(i);
+                      resultRow.put(alias != null ? alias : key, value);
+                    }
+                  }
+                  catch (SQLException e) {
+                    throw new ResultSetException("Unable to access specific metadata from " +
+                                                 "result set metadata", e, ctx);
+                  }
+                  return resultRow;
+                }
+            ).iterator();
+            while (resultIterator.hasNext()) {
+              fos.write(objectMapper.writeValueAsBytes(resultIterator.next()));
+            }
+            return null;
+          },
+          (exception) -> {
+            final boolean isStatementException = exception instanceof StatementException ||
+                                                 (exception instanceof CallbackFailedException
+                                                  && exception.getCause() instanceof StatementException);
+            return sqlMetadataConnector.isTransientException(exception) && !(isStatementException);
```

Comment:
Is it possible that these exceptions can occur in the middle of query execution? If so, the file might contain duplicated rows.

Reply:
Exceptions that occur in the middle of query executions would not be transient exceptions and in those cases, the task will fail.